From 163ae157ea2e2a53e19ca1dc92e6957ea8d0f265 Mon Sep 17 00:00:00 2001
From: Jonas Faber <72028327+Jonohas@users.noreply.github.com>
Date: Mon, 12 Oct 2026 16:15:51 +0200
Subject: [PATCH 06/12] SystemUI: play AOSP fold reveals and Launcher unfold
 only in Pixel

In Stock and Duo the black reveal overlays report ready at once and
never attach, and Launcher3 receives no unfold progress.
---
 .../FoldLightRevealOverlayAnimationTest.kt    |   3 +
 .../UnfoldLightRevealOverlayAnimationTest.kt  | 136 ++++++++++++++++++
 .../yrrp/PixelOnlyProgressListenerTest.kt     |  67 +++++++++
 .../unfold/FoldLightRevealOverlayAnimation.kt |  12 +-
 .../unfold/UnfoldInitializationStartable.kt   |   9 +-
 .../UnfoldLightRevealOverlayAnimation.kt      |  16 ++-
 .../unfold/yrrp/PixelOnlyProgressListener.kt  |  49 +++++++
 7 files changed, 285 insertions(+), 7 deletions(-)
 create mode 100644 packages/SystemUI/multivalentTests/src/com/android/systemui/unfold/UnfoldLightRevealOverlayAnimationTest.kt
 create mode 100644 packages/SystemUI/multivalentTests/src/com/android/systemui/unfold/yrrp/PixelOnlyProgressListenerTest.kt
 create mode 100644 packages/SystemUI/src/com/android/systemui/unfold/yrrp/PixelOnlyProgressListener.kt

diff --git a/packages/SystemUI/multivalentTests/src/com/android/systemui/unfold/FoldLightRevealOverlayAnimationTest.kt b/packages/SystemUI/multivalentTests/src/com/android/systemui/unfold/FoldLightRevealOverlayAnimationTest.kt
index 98d99f781879..3f06d27f9484 100644
--- a/packages/SystemUI/multivalentTests/src/com/android/systemui/unfold/FoldLightRevealOverlayAnimationTest.kt
+++ b/packages/SystemUI/multivalentTests/src/com/android/systemui/unfold/FoldLightRevealOverlayAnimationTest.kt
@@ -36,6 +36,8 @@ import com.android.systemui.power.domain.interactor.powerInteractor
 import com.android.systemui.power.shared.model.ScreenPowerState
 import com.android.systemui.statusbar.LightRevealScrim
 import com.android.systemui.testKosmos
+import com.android.systemui.unfold.yrrp.FoldAnimationStyle
+import com.android.systemui.unfold.yrrp.FoldAnimationStyleSelector
 import com.android.systemui.util.animation.data.repository.fakeAnimationStatusRepository
 import kotlinx.coroutines.test.TestScope
 import kotlinx.coroutines.test.advanceTimeBy
@@ -94,6 +96,7 @@ class FoldLightRevealOverlayAnimationTest : SysuiTestCase() {
                 mockControllerFactory,
                 mockFoldLockSettingAvailabilityProvider,
                 mockJankMonitor,
+                FoldAnimationStyleSelector().apply { onSettingChanged(FoldAnimationStyle.PIXEL) },
             )
         foldLightRevealAnimation.init()
     }
diff --git a/packages/SystemUI/multivalentTests/src/com/android/systemui/unfold/UnfoldLightRevealOverlayAnimationTest.kt b/packages/SystemUI/multivalentTests/src/com/android/systemui/unfold/UnfoldLightRevealOverlayAnimationTest.kt
new file mode 100644
index 000000000000..fa1210f33505
--- /dev/null
+++ b/packages/SystemUI/multivalentTests/src/com/android/systemui/unfold/UnfoldLightRevealOverlayAnimationTest.kt
@@ -0,0 +1,136 @@
+/*
+ * Copyright (C) 2026 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.android.systemui.unfold
+
+import android.hardware.devicestate.DeviceStateManager
+import android.os.Handler
+import android.testing.TestableLooper
+import androidx.test.ext.junit.runners.AndroidJUnit4
+import androidx.test.filters.SmallTest
+import com.android.internal.R
+import com.android.systemui.SysuiTestCase
+import com.android.systemui.flags.FeatureFlagsClassic
+import com.android.systemui.foldedDeviceStateList
+import com.android.systemui.kosmos.Kosmos
+import com.android.systemui.unfold.yrrp.FoldAnimationStyle
+import com.android.systemui.unfold.yrrp.FoldAnimationStyleSelector
+import com.android.systemui.unfoldedDeviceState
+import com.android.systemui.util.concurrency.ThreadFactory
+import java.util.concurrent.Executor
+import javax.inject.Provider
+import org.junit.Before
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.mockito.ArgumentCaptor
+import org.mockito.Mockito.never
+import org.mockito.Mockito.verify
+import org.mockito.kotlin.any
+import org.mockito.kotlin.anyOrNull
+import org.mockito.kotlin.eq
+import org.mockito.kotlin.mock
+import org.mockito.kotlin.whenever
+
+@SmallTest
+@TestableLooper.RunWithLooper
+@RunWith(AndroidJUnit4::class)
+class UnfoldLightRevealOverlayAnimationTest : SysuiTestCase() {
+    private val kosmos = Kosmos()
+    private val foldedState = kosmos.foldedDeviceStateList.first()
+    private val unfoldedState = kosmos.unfoldedDeviceState
+
+    private val selector = FoldAnimationStyleSelector()
+    private val deviceStateManager = mock<DeviceStateManager>()
+    private val threadFactory = mock<ThreadFactory>()
+    private val factory = mock<FullscreenLightRevealAnimationController.Factory>()
+    private val controller = mock<FullscreenLightRevealAnimationController>()
+    private val bgProvider = mock<UnfoldTransitionProgressProvider>()
+    private val mainProvider = mock<UnfoldTransitionProgressProvider>()
+    private val onOverlayReady = mock<Runnable>()
+    private lateinit var testableLooper: TestableLooper
+    private lateinit var foldStateCallback: DeviceStateManager.DeviceStateCallback
+    private lateinit var underTest: UnfoldLightRevealOverlayAnimation
+
+    @Before
+    fun setUp() {
+        testableLooper = TestableLooper.get(this)
+        mContext.orCreateTestableResources.addOverride(
+            R.array.config_foldedDeviceStates,
+            intArrayOf(foldedState.identifier),
+        )
+        whenever(factory.create(any(), any(), any())).thenReturn(controller)
+        whenever(threadFactory.buildDelayableExecutorOnHandler(any())).thenReturn(mock())
+
+        underTest =
+            UnfoldLightRevealOverlayAnimation(
+                mContext,
+                mock<FeatureFlagsClassic>(),
+                mContext.contentResolver,
+                Handler(testableLooper.looper),
+                Provider { bgProvider },
+                Provider { mainProvider },
+                deviceStateManager,
+                threadFactory,
+                factory,
+                selector,
+            )
+        underTest.init()
+
+        val captor = ArgumentCaptor.forClass(DeviceStateManager.DeviceStateCallback::class.java)
+        verify(deviceStateManager).registerCallback(any<Executor>(), captor.capture())
+        foldStateCallback = captor.value
+    }
+
+    @Test
+    fun stock_onScreenTurningOn_reportsReadyWithoutOverlay() {
+        selector.onSettingChanged(FoldAnimationStyle.STOCK)
+        unfoldAfterFold()
+
+        underTest.onScreenTurningOn(onOverlayReady)
+        testableLooper.processAllMessages()
+
+        verify(onOverlayReady).run()
+        verify(controller, never()).addOverlay(any(), anyOrNull())
+    }
+
+    @Test
+    fun duo_onScreenTurningOn_reportsReadyWithoutOverlay() {
+        selector.onSettingChanged(FoldAnimationStyle.DUO)
+        unfoldAfterFold()
+
+        underTest.onScreenTurningOn(onOverlayReady)
+        testableLooper.processAllMessages()
+
+        verify(onOverlayReady).run()
+        verify(controller, never()).addOverlay(any(), anyOrNull())
+    }
+
+    @Test
+    fun pixel_onScreenTurningOn_addsOverlay() {
+        selector.onSettingChanged(FoldAnimationStyle.PIXEL)
+        unfoldAfterFold()
+
+        underTest.onScreenTurningOn(onOverlayReady)
+        testableLooper.processAllMessages()
+
+        verify(controller).addOverlay(any(), eq(onOverlayReady))
+    }
+
+    private fun unfoldAfterFold() {
+        foldStateCallback.onDeviceStateChanged(foldedState)
+        foldStateCallback.onDeviceStateChanged(unfoldedState)
+        testableLooper.processAllMessages()
+    }
+}
diff --git a/packages/SystemUI/multivalentTests/src/com/android/systemui/unfold/yrrp/PixelOnlyProgressListenerTest.kt b/packages/SystemUI/multivalentTests/src/com/android/systemui/unfold/yrrp/PixelOnlyProgressListenerTest.kt
new file mode 100644
index 000000000000..36ab3a2408cb
--- /dev/null
+++ b/packages/SystemUI/multivalentTests/src/com/android/systemui/unfold/yrrp/PixelOnlyProgressListenerTest.kt
@@ -0,0 +1,67 @@
+/*
+ * Copyright (C) 2026 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.android.systemui.unfold.yrrp
+
+import androidx.test.ext.junit.runners.AndroidJUnit4
+import androidx.test.filters.SmallTest
+import com.android.systemui.SysuiTestCase
+import com.android.systemui.unfold.UnfoldTransitionProgressProvider.TransitionProgressListener
+import com.google.common.truth.Truth.assertThat
+import org.junit.Test
+import org.junit.runner.RunWith
+
+@SmallTest
+@RunWith(AndroidJUnit4::class)
+class PixelOnlyProgressListenerTest : SysuiTestCase() {
+    private val selector = FoldAnimationStyleSelector()
+    private val events = mutableListOf<String>()
+    private val delegate =
+        object : TransitionProgressListener {
+            override fun onTransitionStarted() {
+                events += "start"
+            }
+
+            override fun onTransitionProgress(progress: Float) {
+                events += "p"
+            }
+
+            override fun onTransitionFinished() {
+                events += "finish"
+            }
+        }
+    private val underTest = PixelOnlyProgressListener(delegate, selector)
+
+    @Test
+    fun pixel_forwards() {
+        selector.onSettingChanged(FoldAnimationStyle.PIXEL)
+        val t = Any()
+        selector.onGatedTransitionStarted(t)
+        underTest.onTransitionStarted()
+        underTest.onTransitionProgress(0.5f)
+        underTest.onTransitionFinished()
+        assertThat(events).containsExactly("start", "p", "finish").inOrder()
+    }
+
+    @Test
+    fun duo_forwardsNothing() {
+        selector.onSettingChanged(FoldAnimationStyle.DUO)
+        selector.onGatedTransitionStarted(Any())
+        underTest.onTransitionStarted()
+        underTest.onTransitionProgress(0.5f)
+        underTest.onTransitionFinished()
+        assertThat(events).isEmpty()
+    }
+}
diff --git a/packages/SystemUI/src/com/android/systemui/unfold/FoldLightRevealOverlayAnimation.kt b/packages/SystemUI/src/com/android/systemui/unfold/FoldLightRevealOverlayAnimation.kt
index f99094784e63..596f606e7438 100644
--- a/packages/SystemUI/src/com/android/systemui/unfold/FoldLightRevealOverlayAnimation.kt
+++ b/packages/SystemUI/src/com/android/systemui/unfold/FoldLightRevealOverlayAnimation.kt
@@ -37,6 +37,8 @@ import com.android.systemui.unfold.FullscreenLightRevealAnimationController.Comp
 import com.android.systemui.unfold.FullscreenLightRevealAnimationController.Companion.ALPHA_TRANSPARENT
 import com.android.systemui.unfold.FullscreenLightRevealAnimationController.Companion.isVerticalRotation
 import com.android.systemui.unfold.dagger.UnfoldBg
+import com.android.systemui.unfold.yrrp.FoldAnimationStyle
+import com.android.systemui.unfold.yrrp.FoldAnimationStyleSelector
 import com.android.systemui.util.animation.data.repository.AnimationStatusRepository
 import com.android.systemui.util.kotlin.race
 import javax.inject.Inject
@@ -68,7 +70,8 @@ constructor(
     private val animationStatusRepository: AnimationStatusRepository,
     private val controllerFactory: FullscreenLightRevealAnimationController.Factory,
     private val foldLockSettingAvailabilityProvider: FoldLockSettingAvailabilityProvider,
-    private val interactionJankMonitor: InteractionJankMonitor
+    private val interactionJankMonitor: InteractionJankMonitor,
+    private val foldAnimationStyleSelector: FoldAnimationStyleSelector,
 ) : FullscreenLightRevealAnimation {
 
     private val revealProgressValueAnimator: ValueAnimator =
@@ -108,7 +111,12 @@ constructor(
                 .distinctUntilChanged()
                 .flatMapLatest { isFolded ->
                     flow<Nothing> {
-                            if (!areAnimationEnabled.first() || !isFolded) {
+                            if (
+                                !areAnimationEnabled.first() ||
+                                    !isFolded ||
+                                    foldAnimationStyleSelector.currentStyle() !=
+                                        FoldAnimationStyle.PIXEL
+                            ) {
                                 return@flow
                             }
                             race(
diff --git a/packages/SystemUI/src/com/android/systemui/unfold/UnfoldInitializationStartable.kt b/packages/SystemUI/src/com/android/systemui/unfold/UnfoldInitializationStartable.kt
index 67ca2909befe..57c6bdefe330 100644
--- a/packages/SystemUI/src/com/android/systemui/unfold/UnfoldInitializationStartable.kt
+++ b/packages/SystemUI/src/com/android/systemui/unfold/UnfoldInitializationStartable.kt
@@ -20,6 +20,8 @@ import com.android.systemui.CoreStartable
 import com.android.systemui.Flags
 import com.android.systemui.unfold.dagger.UnfoldBg
 import com.android.systemui.unfold.progress.UnfoldTransitionProgressForwarder
+import com.android.systemui.unfold.yrrp.FoldAnimationStyleSelector
+import com.android.systemui.unfold.yrrp.PixelOnlyProgressListener
 import java.util.Optional
 import javax.inject.Inject
 
@@ -34,7 +36,8 @@ constructor(
         Optional<UnfoldTransitionProgressProvider>,
     private val unfoldTransitionProgressProviderOptional:
         Optional<UnfoldTransitionProgressProvider>,
-    private val unfoldTransitionProgressForwarder: Optional<UnfoldTransitionProgressForwarder>
+    private val unfoldTransitionProgressForwarder: Optional<UnfoldTransitionProgressForwarder>,
+    private val foldAnimationStyleSelector: FoldAnimationStyleSelector,
 ) : CoreStartable {
     override fun start() {
         unfoldComponentOptional.ifPresent { c: SysUIUnfoldComponent ->
@@ -59,7 +62,9 @@ constructor(
             progressProvider: UnfoldTransitionProgressProvider ->
             unfoldTransitionProgressForwarder.ifPresent {
                 listener: UnfoldTransitionProgressForwarder ->
-                progressProvider.addCallback(listener)
+                progressProvider.addCallback(
+                    PixelOnlyProgressListener(listener, foldAnimationStyleSelector)
+                )
             }
         }
     }
diff --git a/packages/SystemUI/src/com/android/systemui/unfold/UnfoldLightRevealOverlayAnimation.kt b/packages/SystemUI/src/com/android/systemui/unfold/UnfoldLightRevealOverlayAnimation.kt
index 666e75f20b7f..12a2aee8d486 100644
--- a/packages/SystemUI/src/com/android/systemui/unfold/UnfoldLightRevealOverlayAnimation.kt
+++ b/packages/SystemUI/src/com/android/systemui/unfold/UnfoldLightRevealOverlayAnimation.kt
@@ -33,6 +33,8 @@ import com.android.systemui.unfold.UnfoldLightRevealOverlayAnimation.AddOverlayR
 import com.android.systemui.unfold.UnfoldLightRevealOverlayAnimation.AddOverlayReason.UNFOLD
 import com.android.systemui.unfold.dagger.UnfoldBg
 import com.android.systemui.unfold.util.ScaleAwareTransitionProgressProvider.Companion.areAnimationsEnabled
+import com.android.systemui.unfold.yrrp.FoldAnimationStyle
+import com.android.systemui.unfold.yrrp.FoldAnimationStyleSelector
 import com.android.systemui.util.concurrency.ThreadFactory
 import java.util.concurrent.Executor
 import java.util.function.Consumer
@@ -53,7 +55,8 @@ constructor(
     private val deviceStateManager: DeviceStateManager,
     private val threadFactory: ThreadFactory,
     private val fullscreenLightRevealAnimationControllerFactory:
-        FullscreenLightRevealAnimationController.Factory
+        FullscreenLightRevealAnimationController.Factory,
+    private val foldAnimationStyleSelector: FoldAnimationStyleSelector,
 ) : FullscreenLightRevealAnimation {
 
     private val transitionListener = TransitionListener()
@@ -96,7 +99,12 @@ constructor(
             Trace.beginSection("$TAG#onScreenTurningOn")
             try {
                 // Add the view only if we are unfolding and this is the first screen on
-                if (!isFolded && !isUnfoldHandled && contentResolver.areAnimationsEnabled()) {
+                if (
+                    !isFolded &&
+                        !isUnfoldHandled &&
+                        isPixel() &&
+                        contentResolver.areAnimationsEnabled()
+                ) {
                     overlayAddReason = UNFOLD
                     controller.addOverlay(calculateRevealAmount(), onOverlayReady)
                     isUnfoldHandled = true
@@ -111,6 +119,8 @@ constructor(
         }
     }
 
+    private fun isPixel() = foldAnimationStyleSelector.currentStyle() == FoldAnimationStyle.PIXEL
+
     private fun calculateRevealAmount(animationProgress: Float? = null): Float {
         val overlayAddReason = overlayAddReason
 
@@ -154,7 +164,7 @@ constructor(
 
         override fun onTransitionStarted() {
             // Add view for folding case (when unfolding the view is added earlier)
-            if (controller.isOverlayVisible()) {
+            if (isPixel() && controller.isOverlayVisible()) {
                 executeInBackground {
                     overlayAddReason = FOLD
                     controller.addOverlay(calculateRevealAmount())
diff --git a/packages/SystemUI/src/com/android/systemui/unfold/yrrp/PixelOnlyProgressListener.kt b/packages/SystemUI/src/com/android/systemui/unfold/yrrp/PixelOnlyProgressListener.kt
new file mode 100644
index 000000000000..983028d69a78
--- /dev/null
+++ b/packages/SystemUI/src/com/android/systemui/unfold/yrrp/PixelOnlyProgressListener.kt
@@ -0,0 +1,49 @@
+/*
+ * Copyright (C) 2026 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.android.systemui.unfold.yrrp
+
+import com.android.systemui.unfold.UnfoldTransitionProgressProvider.TransitionProgressListener
+
+/**
+ * Passes a transition to [delegate] only when its latched style is Pixel. Launcher3 receives
+ * progress through this, so its icon motion plays only with the AOSP animation; in Duo Shell
+ * stretches the home task instead.
+ */
+class PixelOnlyProgressListener(
+    private val delegate: TransitionProgressListener,
+    private val selector: FoldAnimationStyleSelector,
+) : TransitionProgressListener {
+    private var passing = false
+
+    override fun onTransitionStarted() {
+        passing = selector.currentStyle() == FoldAnimationStyle.PIXEL
+        if (passing) delegate.onTransitionStarted()
+    }
+
+    override fun onTransitionProgress(progress: Float) {
+        if (passing) delegate.onTransitionProgress(progress)
+    }
+
+    override fun onTransitionFinishing() {
+        if (passing) delegate.onTransitionFinishing()
+    }
+
+    override fun onTransitionFinished() {
+        if (passing) delegate.onTransitionFinished()
+        passing = false
+    }
+}
-- 
2.56.0

